### PR TITLE
Add an Invoke implementation in CodegenDataModel

### PR DESCRIPTION
**Pull request by andy31415**
This removes the previous proposal of using an InvokeResponder and keeps using a CommandHandler. This is because the CommandHandler interface requires a CommandHandler so this is likely the path of least resistance (otherwise defining and implementing an InvokeResponder would likely be quite complex).

After the above changes, the update is quite trivial as the invoke basically translates to a Dispatch() that can never fail. I added a future todo to handle CommandHandlerInterface in a similar way it is handled for attributes, i.e. a separate registry that can be used and accessed. That would allow a DM move flexibility, would make it symmetrical with AAI and likely add more unit testing ability.

### Changes

- Drop all InvokeResponder APIs, replace with CommandHandler
- Make the CommandHandler a separate interface target in src/app/BUILD.gn
- Update invoke signature to be `virtual CHIP_ERROR Invoke(const InvokeRequest & request, chip::TLV::TLVReader & input_arguments, CommandHandler * handler) = 0;`
- Provide an implementation for CodegenDatamodel
- Added some basic unit tests

image

image

### CODE REVIEW

File: src/app/codegen-data-model/CodegenDataModel.cpp [hunk 1]
```diff
@@ -232,10 +233,17 @@ bool CodegenDataModel::EmberCommandListIterator::Exists(const CommandId * list,
 }
 
 CHIP_ERROR CodegenDataModel::Invoke(const InteractionModel::InvokeRequest & request, TLV::TLVReader & input_arguments,
-                                    InteractionModel::InvokeReply & reply)
+                                    CommandHandler * handler)
 {
-    // TODO: this needs an implementation
-    return CHIP_ERROR_NOT_IMPLEMENTED;
+    // TODO: CommandHandlerInterface support is currently
```
Comment (tehampson) [inline src/app/codegen-data-model/CodegenDataModel.cpp:238]: nit, do we have an issue tracker that we can link to. Or is your plan to do this in a follow up really soon?

File: src/app/codegen-data-model/tests/TestCodegenModelViaMocks.cpp [hunk 2]
```diff
@@ -2430,6 +2435,41 @@ TEST(TestCodegenModelViaMocks, EmberWriteAttributeAccessInterfaceTest)
     TestEmberScalarNullWrite<int64_t, ZCL_INT64S_ATTRIBUTE_TYPE>();
 }
 
+TEST(TestCodegenModelViaMocks, EmberInvokeTest)
+{
+    // Ember invoke is extrealy hard-coded, so generally we just need to validate that
```
Comment (tehampson) [inline src/app/codegen-data-model/tests/TestCodegenModelViaMocks.cpp:2440]: nit, what is extrealy hard-coded? Do you mean externally hard-coded?

File: src/app/codegen-data-model/tests/TestCodegenModelViaMocks.cpp [hunk 3]
```diff
@@ -2430,6 +2435,41 @@ TEST(TestCodegenModelViaMocks, EmberWriteAttributeAccessInterfaceTest)
...
+    chip::app::CodegenDataModel model;
+
+    {
+        const ConcreteCommandPath kCommandPath(kMockEndpoint1, MockClusterId(1), kMockCommandId1);
+        const InvokeRequest kInvokeRequest{ .path = kCommandPath };
+        chip::TLV::TLVReader tlvReader;
+
+        const uint32_t kDispatchCountPre = chip::Test::DispatchCount();
+
+        ASSERT_EQ(model.Invoke(kInvokeRequest, tlvReader, nullptr /* handler, NOT used by impl*/), CHIP_NO_ERROR);
```
Comment (tehampson) [inline src/app/codegen-data-model/tests/TestCodegenModelViaMocks.cpp:2453]: nit, Is this coding style allowed? I think I have seem something like this more often
```suggestion
        // Using handler set to nullptr as it is not used by impl
        ASSERT_EQ(model.Invoke(kInvokeRequest, tlvReader, /*handler=*/ nullptr), CHIP_NO_ERROR);
```

File: src/app/codegen-data-model/tests/TestCodegenModelViaMocks.cpp [hunk 4]
```diff
@@ -2430,6 +2435,41 @@ TEST(TestCodegenModelViaMocks, EmberWriteAttributeAccessInterfaceTest)
...
+    }
+
+    {
+        const ConcreteCommandPath kCommandPath(kMockEndpoint1, MockClusterId(1), kMockCommandId2);
+        const InvokeRequest kInvokeRequest{ .path = kCommandPath };
+        chip::TLV::TLVReader tlvReader;
+
+        const uint32_t kDispatchCountPre = chip::Test::DispatchCount();
+
+        ASSERT_EQ(model.Invoke(kInvokeRequest, tlvReader, nullptr /* handler, NOT used by impl*/), CHIP_NO_ERROR);
```
Comment (tehampson) [inline src/app/codegen-data-model/tests/TestCodegenModelViaMocks.cpp:2466]: same

**Review by tehampson — APPROVED** [review]

File: src/app/codegen-data-model/tests/TestCodegenModelViaMocks.cpp (hunk 2 above)
Comment (andy31415) [reply src/app/codegen-data-model/tests/TestCodegenModelViaMocks.cpp:2440]: I think I meant `extreamly`. Re-phrased.

File: src/app/codegen-data-model/tests/TestCodegenModelViaMocks.cpp [hunk 5]
```diff
@@ -2450,7 +2454,8 @@ TEST(TestCodegenModelViaMocks, EmberInvokeTest)
 
         const uint32_t kDispatchCountPre = chip::Test::DispatchCount();
 
-        ASSERT_EQ(model.Invoke(kInvokeRequest, tlvReader, nullptr /* handler, NOT used by impl*/), CHIP_NO_ERROR);
+        // Using a handler set to nullpotr as it is not used by the impl
```
Comment (tehampson) [inline src/app/codegen-data-model/tests/TestCodegenModelViaMocks.cpp:2457]: nit, nullptr

**Review by tehampson — APPROVED** [review]

File: src/app/codegen-data-model/tests/TestCodegenModelViaMocks.cpp (hunk 5 above)
Comment (andy31415) [reply src/app/codegen-data-model/tests/TestCodegenModelViaMocks.cpp:2457]:
```suggestion
        // Using a handler set to nullptr as it is not used by the impl
```

Commenting as suggestion so I can "merge" without checking out the branch myself :)

**Review by bzbarsky-apple — APPROVED** [review]

**Pull request merged**